plugins/backstage-plugin-devcontainers-backend/src/processors/DevcontainersProcessor.test.ts

```diff
@@ -0,0 +1,233 @@
+import {
+  type ReadUrlResponse,
+  type ReadUrlOptions,
+  type SearchResponse,
+  type UrlReader,
+  getVoidLogger,
+} from '@backstage/backend-common';
+import { NotFoundError } from '@backstage/errors';
+import type { LocationSpec } from '@backstage/plugin-catalog-common';
+import {
+  type Entity,
+  ANNOTATION_SOURCE_LOCATION,
+} from '@backstage/catalog-model';
+import {
+  DEFAULT_TAG_NAME,
+  DevcontainersProcessor,
+  PROCESSOR_NAME_PREFIX,
+} from './DevcontainersProcessor';
+
+const mockUrlRoot = 'https://github.com/example-company/example-repo';
+
+const baseEntity: Entity = {
+  apiVersion: 'backstage.io/v1alpha1',
+  kind: 'Component',
+  metadata: {
+    name: 'metadata',
+    tags: [], // Purposefully left empty
+    annotations: {
+      [ANNOTATION_SOURCE_LOCATION]: `${mockUrlRoot}/tree/main`,
+    },
+  },
+};
+
+const baseLocation: LocationSpec = {
+  type: 'Component',
+  presence: 'required',
+  target: `${mockUrlRoot}/blob/main/catalog-info.yaml`,
+};
+
+type SetupOptions = Readonly<{
+  tagName?: string;
+
+  searchThrowCallback?: (
+    url: string,
+    readOptions: ReadUrlOptions | undefined,
+  ) => never;
+
+  readUrlThrowCallback?: (
+    url: string,
+    readOptions: ReadUrlOptions | undefined,
+  ) => never;
+}>;
+
+function setupProcessor(options?: SetupOptions) {
```

Comment:
Could we make the mock take in files so we can test the various locations?  Right now we always return a mock file or always return an error, but if we could do something like:

```
// these get a tag
setupProcessor({
  files: {
    ".devcontainer.json": "blah",
  },
})
setupProcessor({
  files: {
    ".devcontainer/nested/container.json": "blah",
  },
})
// this does not get a tag
setupProcessor({
  files: {
    ".devcontainer/nested/too/much/container.json": "blah",
  },
})
// and so on
```

Then we could test all the permutations (anything not specified returns `NotFoundError`).  Although, the second example will require parsing a glob, but I think we can just import the glob library and compare with each file, which will be perfect so we can make sure we are using the glob correctly as well.

Reply:
Oh, yeah. Good call. I'm going to try fiddling with this, but when you say "glob library", do you mean the one on NPM?

Reply:
Also, just making sure: would `files` be an object, or an array?

Reply:
I think you could do it either way, I went with an object because I think it is slightly simpler on the caller side, but it could be annoying if we need more properties like the file mode or whatever, so array could be more future-proof.

Reply:
Also idk but this might be helpful: https://www.npmjs.com/package/mock-fs

Reply:
Which we can maybe combine with https://github.com/isaacs/node-glob

I think it would let us do something like:

```
mock(files)
```

and then for search we additionally have:

```
await glob(theGlob)
```

or something like that.

Reply:
That makes sense. I can read up on globs and how to handle them – my main concern is because this PR has a fix for the main processor functionality, too, I don't know how long that would delay me wrapping things up

Would you be okay with approving this PR, as long as I make a new issue for tetsting the glob functionality?

Reply:
Yup, no worries.  Hopefully it is not too rough, the `glob` library should handle it all if you mock the filesystem, but reasonable to do it in another PR I think.

Reply:
Cool. Thanks!